internal/lru/lru.go

```diff
@@ -0,0 +1,216 @@
+// after any addition, the least recently used value is evicted.
+func (c *Cache[K, V]) Set(key K, value V) {
+	if c.lookup == nil {
+		c.lookup = make(map[K]*entry[K, V])
+	}
+	if ent, ok := c.lookup[key]; ok {
+		c.moveToFront(ent)
+		ent.value = value
+		return
+	}
+	ent := c.newAtFront(key, value)
+	c.lookup[key] = ent
+	if c.MaxEntries != 0 && c.Len() > c.MaxEntries {
+		c.deleteOldest()
+	}
+}
+
+// Clear removes all items from the cache.
+func (c *Cache[K, V]) Clear() {
+	c.head = nil
+	c.lookup = nil
+}
+
+// Get looks up a key's value from the cache, returning either
+// the value or the zero value if it not present.
+//
+// If found, key is moved to the front of the LRU.
+func (c *Cache[K, V]) Get(key K) V {
+	v, _ := c.GetOk(key)
+	return v
+}
+
+// Contains reports whether c contains key.
+//
+// If found, key is moved to the front of the LRU.
+func (c *Cache[K, V]) Contains(key K) bool {
+	_, ok := c.GetOk(key)
+	return ok
+}
+
+// GetOk looks up a key's value from the cache, also reporting whether
+// it was present.
+//
+// If found, key is moved to the front of the LRU.
+func (c *Cache[K, V]) GetOk(key K) (value V, ok bool) {
+	if ent, hit := c.lookup[key]; hit {
+		c.moveToFront(ent)
+		return ent.value, true
+	}
+	var zero V
+	return zero, false
+}
+
+// PeekOk looks up the key's value from the cache, also reporting
+// whether it was present.
+//
+// Unlike GetOk, PeekOk does not move key to the front of the
+// LRU. This should mostly be used for non-intrusive debug inspection
+// of the cache.
+func (c *Cache[K, V]) PeekOk(key K) (value V, ok bool) {
+	if ent, hit := c.lookup[key]; hit {
+		return ent.value, true
+	}
+	var zero V
+	return zero, false
+}
+
+// Delete removes the provided key from the cache if it was present.
+func (c *Cache[K, V]) Delete(key K) {
+	if ent, ok := c.lookup[key]; ok {
+		c.deleteElement(ent)
+	}
+}
+
+// DeleteOldest removes the item from the cache that was least
+// recently accessed. It is a no-op if the cache is empty.
+func (c *Cache[K, V]) DeleteOldest() {
+	if c.head != nil {
+		c.deleteOldest()
+	}
+}
+
+// Len returns the number of items in the cache.
+func (c *Cache[K, V]) Len() int { return len(c.lookup) }
+
+// newAtFront creates a new LRU entry using key and value, and inserts
+// it at the front of c.head.
+func (c *Cache[K, V]) newAtFront(key K, value V) *entry[K, V] {
+	ret := &entry[K, V]{key: key, value: value}
+	if c.head == nil {
+		ret.prev = ret
+		ret.next = ret
+	} else {
+		ret.next = c.head
+		ret.prev = c.head.prev
+		c.head.prev.next = ret
+		c.head.prev = ret
+	}
+	c.head = ret
+	return ret
+}
+
+// moveToFront moves ent, which must be an existing element of the
+// cache, to the front of c.head.
+func (c *Cache[K, V]) moveToFront(ent *entry[K, V]) {
+	if c.head == ent {
+		return
+	}
+	ent.prev.next = ent.next
+	ent.next.prev = ent.prev
+	ent.prev = c.head.prev
+	ent.next = c.head
+	c.head.prev.next = ent
+	c.head.prev = ent
+	c.head = ent
```

Comment:
Do we not need to lock the cache?

Reply:
It's documented as unlocked. BYOLocking.